File: packages/unity-editor/Editor/Matcher/Tokenize.cs
using System.Collections.Generic;
using System.Linq;
using System.Text.RegularExpressions;

namespace UiAssemblerSlice.Editor.Matcher
{
    // Ports matcher/tokenize.ts verbatim - shared token-similarity
    // primitives for Candidates' coarse retrieval and StructuralSignal's
    // per-candidate name/description signal.
    internal static class Tokenize
    {
        private static readonly Regex SeparatorPattern = new Regex(@"[_\-./]+", RegexOptions.Compiled);
        private static readonly Regex CamelBoundaryPattern = new Regex(@"([a-z0-9])([A-Z])", RegexOptions.Compiled);
        private static readonly char[] WhitespaceChars = { ' ', '\t', '\n', '\r' };

        // Splits snake_case, kebab-case, path separators, and
        // camelCase/PascalCase into lowercase tokens, e.g.
        // "UIElements__button_green" -> {uielements, button, green}.
        public static HashSet<string> Tokens(string text)
        {
            var spaced = SeparatorPattern.Replace(text ?? "", " ");
            spaced = CamelBoundaryPattern.Replace(spaced, "$1 $2");
            return new HashSet<string>(
                spaced.ToLowerInvariant()
                    .Split(WhitespaceChars, System.StringSplitOptions.RemoveEmptyEntries));
        }

        public static double JaccardSimilarity(HashSet<string> a, HashSet<string> b)
        {
            if (a.Count == 0 || b.Count == 0) return 0;
            var intersection = a.Count(b.Contains);
            var union = a.Count + b.Count - intersection;
            return union == 0 ? 0 : (double)intersection / union;
        }
    }
}

File: packages/mcp-tool/src/matcher/tokenize.ts
// Shared token-similarity primitives for the matcher. candidates.ts's coarse
// retrieval and structural-signal.ts's per-candidate name/description signal
// both need the same "split identifiers into comparable words" step.

// Splits snake_case, kebab-case, path separators, and camelCase/PascalCase
// into lowercase tokens, e.g. "UIElements__button_green" -> {uielements,
// button, green}.
export function tokenize(text: string): Set<string> {
  const spaced = text
    .replace(/[_\-./]+/g, " ")
    .replace(/([a-z0-9])([A-Z])/g, "$1 $2");
  return new Set(
    spaced
      .toLowerCase()
      .split(/\s+/)
      .filter((token) => token.length > 0),
  );
}

export function jaccardSimilarity(a: Set<string>, b: Set<string>): number {
  if (a.size === 0 || b.size === 0) return 0;
  let intersection = 0;
  for (const token of a) {
    if (b.has(token)) intersection++;
  }
  const union = a.size + b.size - intersection;
  return union === 0 ? 0 : intersection / union;
}
